### DISCUSSION TITLE
"relation does not exist" / "no healthy nodes" / "zmq4: could not dial" / "i/o timeout"

### DISCUSSION THREAD

**Discussion by BenzMo** [dreams-money/merged-mining-pool#9] [General]
Hi, so I'm trying to fix this problem for a while now. I've got the Dogecoin and Litecoin nodes running successfully, rpc all works as tested via command line etc. I try to start the pool and get the error below.

root@related-urchin:~/merged-mining-pool# go run main.go
2024/06/11 21:03:51 Started Pool on port: 3643
2024/06/11 21:03:51 Stat Manager running every 2m0s with a hashrate window of 10m0s
2024/06/11 21:03:51 Started API on port: 8001
2024/06/11 21:03:51 Payouts manager running every 10m0s
STATS START
2024/06/11 21:03:51 Total CPU cores avaliable: 16
2024/06/11 21:03:51 Total Goroutines: 6
2024/06/11 21:03:51 Total System Memory: 7445520
2024/06/11 21:03:51 Total Memory Allocated: 335664
STATS END
2024/06/11 21:03:51 Share buffer flushes every 5s
2024/06/11 21:03:56 pq: relation "shares" does not exist
panic: litecoin
no healthy litecoin nodes!

goroutine 20 [running]:
designs.capital/dogepool/pool.panicOnError(...)
        /root/merged-mining-pool/pool/server.go:114
designs.capital/dogepool/pool.(*PoolServer).Start(0xc000156d80)
        /root/merged-mining-pool/pool/server.go:55 +0x1cb
created by main.startPoolServer in goroutine 1
        /root/merged-mining-pool/main.go:45 +0x5f
exit status 2
root@related-urchin:~/merged-mining-pool# sudo -u postgres psql
psql (16.3 (Ubuntu 16.3-1.pgdg22.04+1))
Type "help" for help.

I figured maybe it's an issue with db, so I tried everything I could find online re Postgres relationship errors like adjusting the schema pathing, adjusting permissions, checking connection but nothing is working. On command line I try to access the db 'mergedmining' but it still shows no relations but weirdly when I run /dt when I'm not connected the db directly I get a correct list of relations (see below for example).

I then started a new fresh instance to reinstall and note down my steps (see below)

Postgres

sudo apt update && sudo apt upgrade
sudo sh -c 'echo "deb https://apt.PostgreSQL.org/pub/repos/apt $(lsb_release -cs)-pgdg main" > /etc/apt/sources.list.d/pgdg.list'
wget --quiet -O - https://www.PostgreSQL.org/media/keys/ACCC4CF8.asc | sudo apt-key add -
sudo apt update
sudo apt install postgresql-16 -y
sudo systemctl status  postgresql
cd
git clone https://github.com/dreams-money/merged-mining-pool.git
cd merged-mining-pool/persistence/schema
// I then copied the .sql scripts to /var/lib/postgresql so that I had read permission.
sudo -i -u postgres
psql
\i 1-initiate.sql
\i 2-schema.sql
exit
psql
\i 3-multi-pool-partition.sql
\q
psql
\dt
// Returns
                      List of relations
 Schema |      Name       |       Type        |    Owner
--------+-----------------+-------------------+--------------
 public | balance_changes | table             | mergedmining
 public | balances        | table             | mergedmining
 public | blocks          | table             | mergedmining
 public | miner_settings  | table             | mergedmining
 public | minerstats      | table             | mergedmining
 public | payments        | table             | mergedmining
 public | poolstats       | table             | mergedmining
 public | shares          | partitioned table | mergedmining
 public | shares_mypool1  | table             | mergedmining
(9 rows)
\\
\c mergedmining
\\ You are now connected to database "mergedmining" as user "postgres".
\dt
\\ Did not find any relations.
\c "host=127.0.0.1 port=5432 dbname=mergedmining user=mergedmining password=asdfasdf"
\\ SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, compression: off)
\\ You are now connected to database "mergedmining" as user "mergedmining" on host "127.0.0.1" at port "5432".
\dt
\\ Did not find any relations.

So still getting the same and I expect if I were to install the nodes and run go on this instance I would get the same issue.

Any ideas please? I am stuck. If you provide an ETH address happy to provide a small bounty for a successful fix, Sorry for poor formatting above, not used to Github.

Thanks!

**Comment by BenzMo** [comment]
Update: Okay I think I figured it out (may be obvious to some) but I think the initial script wasn't creating the tables in the mergedmining database due to the commented out /c line. I entered the lines manually and got past the relation error. **However** I'm still stuck on the panic Litecoin error:

STATS END
2024/06/12 04:53:04 Share buffer flushes every 5s
panic: litecoin
no healthy litecoin nodes!

goroutine 20 [running]:
designs.capital/dogepool/pool.panicOnError(...)
        /root/merged-mining-pool/pool/server.go:114
designs.capital/dogepool/pool.(*PoolServer).Start(0xc000156d80)
        /root/merged-mining-pool/pool/server.go:55 +0x1cb
created by main.startPoolServer in goroutine 1
        /root/merged-mining-pool/main.go:45 +0x5f
exit status 2

Hoping I can figure this one out :D

**Reply by BenzMo** [reply]
Update: Okay so I have a theory, I think in health.go it determines if a node has an error and therefore not healthy by calling getblocktemplate, I tried running getblocktemplate in dogecoin-cli and litecoin-cli and got errors because I haven't fully synced/downloaded the full chain yet. I'm hoping when sync is complete the error will go away and it will work. Fingers crossed!

**Comment by BenzMo** [comment]
Update: Okay so healthy node error is now gone as blockchains are synced. However new error to figure out:

STATS END
2024/06/12 18:40:10 Share buffer flushes every 5s
2024/06/12 18:40:10 Sent work to 0 client(s)
panic: zmq4: could not dial to "tcp://localhost:1222" (retry=250ms): dial tcp 127.0.0.1:1222: connect: connection refused

goroutine 7 [running]:
designs.capital/dogepool/pool.panicOnError(...)
        /root/merged-mining-pool/pool/server.go:114
designs.capital/dogepool/pool.(*PoolServer).Start(0xc0000c0d80)
        /root/merged-mining-pool/pool/server.go:63 +0x1be
created by main.startPoolServer in goroutine 1
        /root/merged-mining-pool/main.go:45 +0x5f
exit status 2

**Comment by BenzMo** [comment]
Update: Okay so past the zmq one by adding the following to conf files (editing ports for Litecoin)

zmqpubrawblock=tcp://127.0.0.1:1220
zmqpubrawtx=tcp://127.0.0.1:1221
zmqpubhashblock=tcp://127.0.0.1:1222

Feel that that should have been obvious so my bad. ANDD.. it works server is up and running, I used screen to keep it running in background.

Next problem.. I tried to connect a miner I rented from MMR and I'm getting this message on console:

2024/06/12 21:59:55 Successfully saved stats
2024/06/12 21:59:56 New Stratum Connection from: _miner ip address removed_
2024/06/12 21:59:56 New subscription from client type: GodMiner/1.0.0
2024/06/12 21:59:56 Authorized rig: 123459 mining to addresses: [_ltc address removed_ _doge address removed_]
2024/06/12 22:00:06 Socket read error from: _miner ip address removed_
2024/06/12 22:00:06 read tcp _server ip address removed_:3643->_miner ip address removed_:_miner port removed_: i/o timeout
2024/06/12 22:00:23 New dogecoin block: 118 - e2d484a8cbcbae4b4d67a897f0170169802b173e65092fec638f64cac2c91d23
2024/06/12 22:00:23 Sent work to 0 client(s)
2024/06/12 22:00:36 New Stratum Connection from: _miner ip address removed_
2024/06/12 22:00:36 New subscription from client type: GodMiner/1.0.0
2024/06/12 22:00:36 Authorized rig: 123459 mining to addresses: [_ltc address removed_ _doge address removed_]
2024/06/12 22:00:46 Socket read error from: _miner ip address removed_
2024/06/12 22:00:46 read tcp _server ip address removed_:3643->_miner ip address removed_:_miner port removed_: i/o timeout

I think I'm going to struggle with this one though.

**Reply by BenzMo** [reply]
Update: I changed ports as an NMAP scan was showing this on port 3643

PORT     STATE SERVICE
3643/tcp open  audiojuggler

However it was still showing the same error. I forgot to say initially it also included the following error on first connection:
2024/06/13 04:01:34 Malformed stratum request from: _miner ip address removed_
2024/06/13 04:01:34 invalid character '\x16' looking for beginning of value

When I googled it, I think it was saying it's normally related to a misformatted json string, so I thought maybe it was client specific, I then tried a different miner that was running CGMiner and not Godminer and it's working!

**Reply by BenzMo** [reply]
Update I'm still getting a-lot of socket read errors and i/o timeouts. I don't know if that's normal though.

**Reply by dreams-money** [reply]
Update the timeout in the config file to see if it helps.

**Reply by BenzMo** [reply]
Thanks for the suggestion, I changed it to 600s and it's working! Letting it run for a bit but I think it's doing expected hash rate now.
